controllers/catalogsource_controller.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/operator-framework/operator-controller/internal/resolution/entity_sources/catalogsource"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	defaultCatalogSourceSyncInterval     = 5 * time.Minute
+	defaultRegistryGRPCConnectionTimeout = 10 * time.Second
+
+	eventTypeNormal  = "Normal"
+	eventTypeWarning = "Warning"
+
+	eventReasonCacheUpdated      = "BundleCacheUpdated"
+	eventReasonCacheUpdateFailed = "BundleCacheUpdateFailed"
+)
+
+type CatalogSourceReconcilerOption func(reconciler *CatalogSourceReconciler)
+
+func WithRegistryClient(registry catalogsource.RegistryClient) CatalogSourceReconcilerOption {
+	return func(reconciler *CatalogSourceReconciler) {
+		reconciler.registry = registry
+	}
+}
+
+// applyDefaults applies default values to empty CatalogSourceReconciler fields _after_ options have been applied
+func applyDefaults() CatalogSourceReconcilerOption {
+	return func(reconciler *CatalogSourceReconciler) {
+		if reconciler.registry == nil {
+			reconciler.registry = catalogsource.NewRegistryGRPCClient(defaultRegistryGRPCConnectionTimeout)
+		}
+	}
+}
+
+type CatalogSourceReconciler struct {
+	sync.RWMutex
+	client.Client
+	scheme   *runtime.Scheme
+	registry catalogsource.RegistryClient
+	recorder record.EventRecorder
+	cache    map[string]map[deppy.Identifier]*input.Entity
+}
+
+func NewCatalogSourceReconciler(client client.Client, scheme *runtime.Scheme, recorder record.EventRecorder, options ...CatalogSourceReconcilerOption) *CatalogSourceReconciler {
+	reconciler := &CatalogSourceReconciler{
+		RWMutex:  sync.RWMutex{},
+		Client:   client,
+		scheme:   scheme,
+		recorder: recorder,
+		cache:    map[string]map[deppy.Identifier]*input.Entity{},
+	}
+	// apply options
+	options = append(options, applyDefaults())
+	for _, option := range options {
+		option(reconciler)
+	}
+
+	return reconciler
+}
+
+// +kubebuilder:rbac:groups=operators.coreos.com,resources=catalogsources,verbs=get;list;watch
+
+func (r *CatalogSourceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithName("catalogsource-controller")
+	l.V(1).Info("starting")
+	defer l.V(1).Info("ending")
+
+	var catalogSource = &v1alpha1.CatalogSource{}
+	if err := r.Client.Get(ctx, req.NamespacedName, catalogSource); err != nil {
+		if errors.IsNotFound(err) {
+			r.dropSource(req.String())
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	entities, err := r.registry.ListEntities(ctx, catalogSource)
+	if err != nil {
+		r.recorder.Event(catalogSource, eventTypeWarning, eventReasonCacheUpdateFailed, fmt.Sprintf("Failed to update bundle cache from %s/%s: %v", catalogSource.GetNamespace(), catalogSource.GetName(), err))
+		return ctrl.Result{Requeue: true}, err
```

Comment:
Sort of inline with my previous comment about checking certain aspects of the catsrc status before attempting to ListEntities... Do we need `Requeue: true` here? This will result in an exponential retry even when the catsrc object doesn't change.

I imagine that when a catsrc is coming up, it'll be unavailable until the pod spins up and the grpc server starts. In all that time, we'll be retrying over and over unnecessarily.

It seems like we could wait for cat src events to come into our watch, look at the status, and be a bit more diligent about waiting to attempt ListEntities.

WDYT?

Reply:
Bummer. It would be great if the catalog operator could be the thing probing the underlying GRPC server and telling us via catsrc status updates when it is ready.

Reply:
I thought this information (if the registry server is ready or not) is already exposed via the [GrpcConnectionState](https://github.com/operator-framework/api/blob/master/pkg/operators/v1alpha1/catalogsource_types.go#L191-L195) object in the status?

Reply:
> I thought this information (if the registry server is ready or not) is already exposed via the [GrpcConnectionState](https://github.com/operator-framework/api/blob/master/pkg/operators/v1alpha1/catalogsource_types.go#L191-L195) object in the status?

@anik120 connection state _is_ exposed by GrpcConnectionState in catalogSources when the OLMv0 catalog-operator is running on cluster. On re-checking the code, it looks like it does this for static address catalogSources as well.

Having this as a strict requirement would mean we can't run operator-controller without running OLMv0 catalog-operator on cluster, though with the current reconciler, it'd still be relying on something external to ensure that the catalogSource was being served at the address it points to.